backend/src/controllers/regionController.js

```diff
@@ -1,56 +1,146 @@
-    const regionId  = req.params.regionId;
-    const resolveEmpty = req.query.resolveEmpty === 'true';
-    const hierarchyId = req.query.hierarchyId || 1;
-
+// Retrieve subregions for a specific region
+async function getSubregions(regionId, hierarchyId, getAll) {
+  try {
     // Check if the region exists
     const region = await Hierarchy.findOne({
-        where: {
-            regionId: regionId,
-            hierarchyId: hierarchyId
-        },
-    })
+      where: {
+        regionId,
+        hierarchyId,
+      },
+    });
+
     if (!region) {
-        return res.status(404).json({ message: 'Region not found' });
+      return { data: [], message: 'Region not found', status: 404 };
+    }
+
+    // Retrieve subregions
+    let subregions;
+    // Check the getAll query parameter
+    if (getAll === 'true') {
+      subregions = await getAllSubregions(regionId, hierarchyId);
+    } else {
+      subregions = await Hierarchy.findAll({
+        where: { parentId: regionId, hierarchyId },
+        mapToModel: true,
+        model: Hierarchy,
+      });
+    }
+
+    if (subregions.length === 0) {
+      return { data: [], message: 'Region has no subregions', status: 202 };
     }
 
-    //Find all subdivisions of the region
-    const divisions = await getDivisions(regionId, hierarchyId, res);
-    let geometries = [];
-    let notCompleted = false;
-    for (let division of divisions) {
-        const regionId = division.id;
-        const geom = division.geom;
-        if (geom) {
-            geometries.push(geom);
-        } else {
-            if (!resolveEmpty) {
-                notCompleted = true;
-                break;
-            } else {
-                // Find ids and geometries of all subdivisions of the division
-                const query = `
+    return { data: subregions, status: 200 };
+  } catch (err) {
+    console.error(err);
+    return { data: [], message: 'Internal Server Error', status: 500 };
+  }
+}
+
+// Retrieve the divisions of a region. It does not include subdivisions of the divisions.
+async function getDivisions(regionId, hierarchyId) {
+  const regions = (await getSubregions(regionId, hierarchyId, false)).data;
+  // Add the region itself
+  regions.push(await Hierarchy.findOne({
+    where: {
+      regionId,
+      hierarchyId,
+    },
+  }));
+  let resultDivisions = [];
+  const promises = regions.map(async (region) => {
+    const query = `
+            SELECT r.* FROM hierarchy_region_mapping hrm
+            JOIN regions r ON hrm.region_id = r.id
+            WHERE alt_region_id = :regionId AND hierarchy_id = :hierarchyId
+            `;
+    return sequelize.query(query, {
+      replacements: { regionId: region.regionId, hierarchyId },
+      type: QueryTypes.SELECT,
+      mapToModel: true,
+      model: Region,
+    });
+  });
+  const results = await Promise.all(promises);
+  results.forEach((result) => {
+    resultDivisions = resultDivisions.concat(result.map((r) => r.dataValues));
```

Comment:
**Suggestion:** Use a more descriptive variable name than 'r' for better readability.
```suggestion
    resultDivisions = resultDivisions.concat(result.map((region) => region.dataValues));
```